src/uct/ib/rc/accel/rc_mlx5_iface.c

```diff
@@ -162,10 +162,39 @@ uct_rc_mlx5_iface_progress_common(uct_rc_mlx5_iface_common_t *iface,
     return 0;
 }
 
+static UCS_F_NOINLINE unsigned
+uct_rc_mlx5_common_ka_progress(uct_rc_mlx5_iface_common_t *iface)
+{
+    uct_rc_mlx5_ep_t *ep;
+
+    if (ucs_unlikely(ucs_get_time() < iface->keepalive.time)) {
+        return 0;
+    }
+
+    ucs_spin_lock(&iface->super.ep_list_lock);
+    ucs_list_for_each(ep, &iface->super.ep_list, super.list) {
+        ucs_trace("send keepalive grant on ep %p", ep);
+        uct_rc_ep_fc_send_grant(&ep->super);
+    }
+    ucs_spin_unlock(&iface->super.ep_list_lock);
+
+    uct_rc_mlx5_iface_print(iface, "keepalive");
+
+    iface->keepalive.time = ucs_get_time() + iface->config.ka_interval;
+
+    return 1;
+}
+
 unsigned uct_rc_mlx5_iface_progress(void *arg)
 {
     uct_rc_mlx5_iface_common_t *iface = arg;
 
+    if ((iface->config.ka_interval != 0) &&
+        ((iface->keepalive.iter_count++ %
+          UCP_WORKER_KEEPALIVE_ITER_SKIP) != 0)) {
```

Comment:
> minor: can invert condition order. since we always enable keepalive, probably the condition with % will be false most of the times

I set such condition sequence to not update iter_count in case if KA is not enabled...
updated